File: verdigraph/evaluation.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List


@dataclass(frozen=True)
class EvaluationResult:
    """Outcome signal used as selective pressure for growth/pruning."""

    task_id: str
    task_type: str
    success_score: float
    accuracy: float = 0.0
    user_satisfaction: float = 0.0
    cost_efficiency: float = 0.0
    safety_score: float = 1.0
    notes: str = ""
    used_edges: List[tuple[str, str]] = field(default_factory=list)
    used_nodes: List[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        for name in ("success_score", "accuracy", "user_satisfaction", "cost_efficiency", "safety_score"):
            value = getattr(self, name)
            if not 0.0 <= value <= 1.0:
                raise ValueError(f"{name} must be in [0, 1]")

    @property
    def is_success(self) -> bool:
        return self.success_score >= 0.65 and self.safety_score >= 0.8

    @property
    def is_failure(self) -> bool:
        return self.success_score < 0.4 or self.safety_score < 0.6


class SimpleEvaluator:
    """Deterministic evaluator for demos/tests.

    A real implementation can replace this with task-specific tests, human feedback,
    LLM-as-judge with guardrails, or benchmark scoring.
    """

    def evaluate(self, task_id: str, task_type: str, score: float, used_edges: List[tuple[str, str]], used_nodes: List[str]) -> EvaluationResult:
        return EvaluationResult(
            task_id=task_id,
            task_type=task_type,
            success_score=score,
            accuracy=score,
            user_satisfaction=score,
            cost_efficiency=max(0.0, 1.0 - len(used_nodes) * 0.05),
            safety_score=1.0,
            used_edges=used_edges,
            used_nodes=used_nodes,
            notes="Synthetic demo evaluation.",
        )

File: hosted-mcp/src/verdigraph/evaluation.ts
// src/verdigraph/evaluation.ts — TS port of verdigraph/evaluation.py.

export interface EvaluationResult {
  task_id: string;
  task_type: string;
  success_score: number;
  accuracy: number;
  user_satisfaction: number;
  cost_efficiency: number;
  safety_score: number;
  notes: string;
  used_edges: Array<[string, string]>;
  used_nodes: string[];
}

const inUnitRange = (name: string, v: number) => {
  if (!(v >= 0 && v <= 1)) throw new Error(`${name} must be in [0, 1] (got ${v})`);
};

export function makeEvaluationResult(input: Partial<EvaluationResult> & { task_id: string; task_type: string; success_score: number }): EvaluationResult {
  const r: EvaluationResult = {
    task_id:           input.task_id,
    task_type:         input.task_type,
    success_score:     input.success_score,
    accuracy:          input.accuracy          ?? 0,
    user_satisfaction: input.user_satisfaction ?? 0,
    cost_efficiency:   input.cost_efficiency   ?? 0,
    safety_score:      input.safety_score      ?? 1,
    notes:             input.notes             ?? "",
    used_edges:        input.used_edges        ?? [],
    used_nodes:        input.used_nodes        ?? [],
  };
  inUnitRange("success_score", r.success_score);
  inUnitRange("accuracy", r.accuracy);
  inUnitRange("user_satisfaction", r.user_satisfaction);
  inUnitRange("cost_efficiency", r.cost_efficiency);
  inUnitRange("safety_score", r.safety_score);
  return r;
}

export function isSuccess(r: EvaluationResult): boolean {
  return r.success_score >= 0.65 && r.safety_score >= 0.8;
}

export function isFailure(r: EvaluationResult): boolean {
  return r.success_score < 0.4 || r.safety_score < 0.6;
}
